proposals/infra/INF-0005-register-types-and-diagnostics.md

```diff
@@ -130,21 +130,275 @@ is no corresponding member, a warning will be emitted that is treated as an erro
+The compiler will emit diagnostics by iterating over all Decl objects with a `register` annotation,
+and for each annotation, running `DiagnoseHLSLRegisterAttribute` once.
+
+All the warnings introduced in this spec were not emitted in legacy versions of the compiler.
+The warnings, `warn_hlsl_register_type_c_packoffset`,
+`warn_hlsl_deprecated_register_type_b`,
+`warn_hlsl_deprecated_register_type_i`, and
+`warn_hlsl_user_defined_type_missing_member`, are all within a warning group known as
+`DisallowLegacyBindingRules`. However, only `warn_hlsl_user_defined_type_missing_member` is not 
+treated as errors by default, the rest of the warnings are errors by default. If legacy behavior 
+is desired, a user can pass the `-Wno-disallow-legacy-binding-rules` flag to the compiler to silence 
+the errors and warnings.
+
+
+Here is some pseudocode summarizing the diagnostic emission process:
+
+```
+if the register type is invalid:
+  emit err_hlsl_binding_type_invalid
+
+if Other is set:
+  emit err_hlsl_binding_type_mismatch
+
+if multiple register annotations exist:
+  if any pair of register annotations share the same register type:
+    emit err_hlsl_duplicate_register_annotation
+
+if Resource is set:
+  if register type does not match resource class flag:
+    emit err_hlsl_binding_type_mismatch
+
+if Basic is set:
+  if DefaultGlobals is set:
+    if register type is 'b' or 'i':
+      emit warn_hlsl_deprecated_register_type_b or warn_hlsl_deprecated_register_type_i respectively
+
+  if register type is 'c'
+    if DefaultGlobals is not set:
+      emit warn_hlsl_register_type_c_packoffset
+  else if register type is 't' 'u' or 's':
+    emit err_hlsl_binding_type_mismatch
+  else: emit err_hlsl_binding_type_invalid
+
+if UDT is set:
+  if register type is t:
+    if SRV is not set:
+      emit warn_hlsl_user_defined_type_missing_member
+  else if register type is u:
+    if UAV is not set:
+      emit warn_hlsl_user_defined_type_missing_member
+  else if register type is b:
+    if CBV is not set:
+      emit warn_hlsl_user_defined_type_missing_member
+  else if register type is s:
+    if Sampler is not set:
+      emit warn_hlsl_user_defined_type_missing_member
+  else if register type is c:
+    if ContainsNumeric is not set:
+      emit warn_hlsl_user_defined_type_missing_member
+  else
+    emit err_hlsl_binding_type_invalid  
+```
+
+## Behavioral Differences
+
+This infrastructure will introduce some behavioral differences between `clang` and `dxc`.
+The whole approach of setting flags based on decl characteristics and using them
+to drive diagnostics is an approach that wasn't taken in DXC. Secondly, as mentioned above, 
+the `disallow-legacy-binding-rules` warning group did not exist in `dxc`, and neither
+did any of the warnings that are contained in that group. Those warnings are being 
+introduced to `clang-dxc`. Many of these warnings will be treated as errors, causing some

```

Comment:
nit: "clang-dxc" is the driver.  We're introducing these new warnings and errors to clang. It's possible to light up all this proposed functionality without going through clang-dxc.